File: src/main/java/com/nuutrai/reactor/util/MultiTypeMap.java
package com.nuutrai.reactor.util;

import com.google.common.collect.Maps;
import org.apache.commons.lang3.SerializationUtils;

import java.io.Serializable;
import java.util.Map;

public class MultiTypeMap {

    private final Map<String, byte[]> map = Maps.newHashMap();

    public boolean add(String key, Object value) {
        if (value instanceof Serializable s) {
            map.put(key, SerializationUtils.serialize(s));
            return true;
        }
        return false;
    }

    @SuppressWarnings("unchecked")
    public <T extends Serializable> T get(String key, Class<? extends T> clazz) {
        Object value = SerializationUtils.deserialize(map.get(key));

        if (clazz.isInstance(value)) {
            return (T) value;
        }

        return null;
    }

}

File: src/main/kotlin/com/nuutrai/reactor/util/MultiTypeMap.kt
@file:Suppress("UNCHECKED_CAST")

package com.nuutrai.reactor.util

import org.apache.commons.lang3.SerializationUtils
import java.io.Serializable
import kotlin.reflect.KClass

class MultiTypeMap {
	val map: MutableMap<String, ByteArray> = mutableMapOf()
		private set

	fun add(key: String, value: Any): Boolean {
		if (value is Serializable) {
			map.put(key, SerializationUtils.serialize(value))
			return true
		}
		return false
	}

	inline operator fun <reified T : Serializable> get(key: String): T {
		val value = SerializationUtils.deserialize<T>(map[key])

		if (value == null) {
			throw IllegalArgumentException("Object from MultiTypeMap is null: $key")
		}
		if (T::class.isInstance(value)) {
			return value
		}
		throw IllegalArgumentException("Object from MultiTypeMap does not exist: $key")
	}

	inline operator fun <reified T : Serializable> get(key: String, clazz: KClass<out T>): T {
		val value = SerializationUtils.deserialize<T>(map[key])

		if (value == null) {
			throw IllegalArgumentException("Object from MultiTypeMap is null: $key")
		}
		if (clazz.isInstance(value)) {
			return value
		}
		throw IllegalArgumentException("Object from MultiTypeMap does not exist: $key")
	}

}
